### PR TITLE
SAA-1528 - subject-access-request base allocation searches on allocation time for the from and to date

### PR DESCRIPTION
**Pull request by dafhoustonmoj**
API endpoint GET /subject-access-request/prisoner/{prisonerNumber}

change the from and to date search for allocations to be based on the allocation creation date.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SubjectAccessRequestService.kt [hunk 1]
```diff
@@ -28,7 +28,7 @@ class SubjectAccessRequestService(private val repository: SarRepository) {
     val from = fromDate ?: LocalDate.now()
     val to = toDate ?: LocalDate.now()
 
-    val allocations = repository.findAllocationsBy(prisonerNumber, from, to)
+    val allocations = repository.findAllocationsBy(prisonerNumber, from, to.plusDays(1))
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SubjectAccessRequestService.kt:31]: Given we have to adjust the `to` date on each call I'd be tempted to either just do it once above or even push/hide that logic of altering the date into the SarRepository?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/SarRepository.kt [hunk 2]
```diff
@@ -10,14 +9,7 @@ import java.time.LocalDate
...
 interface SarAllocationRepository : ReadOnlyRepository<SarAllocation, Long> {
-  @Query(
-    """
-    SELECT sa FROM SarAllocation sa
-     WHERE sa.prisonerNumber = :prisonerNumber
-       AND sa.startDate >= :fromDate AND sa.startDate <= :toDate
-    """,
-  )
-  fun findBy(@Param("prisonerNumber") prisonerNumber: String, @Param("fromDate") fromDat: LocalDate, @Param("toDate") toDate: LocalDate): List<SarAllocation>
+  fun findByPrisonerNumberAndCreatedDateBetween(@Param("prisonerNumber") prisonerNumber: String, @Param("fromDate") fromDat: LocalDate, @Param("toDate") toDate: LocalDate): List<SarAllocation>
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/SarRepository.kt:12]: Just noticed we can probably do away with the `@Param(...)` not needed.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM a few minor comments.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SubjectAccessRequestService.kt (hunk 1 above)
Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/SubjectAccessRequestService.kt:31]: I'll look at this. makes sense

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/SarRepository.kt (hunk 2 above)
Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/repository/SarRepository.kt:12]: will check

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**